### PR TITLE
Remove unnecessary restriction to `StridedVecOrMat`

### PR DESCRIPTION
**Pull request by dlfivefifty**
The "Strided array interface" https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-strided-arrays-1 means that this is useful beyond just `StridedVecOrMat` so there is no reason to be so restrictive.

### CODE REVIEW

**Comment by dlfivefifty** [comment]
Note ArrayLayouts.jl also has an approach for pointers to adjoints of complex arrays by defining `ConjPtr` so that, for example, `pointer(view(A',1:3,2:4)')` works. This could be moved to Base and thereby removing type piracy for ArrayLayouts.jl:
```julia
"""
    ConjPtr{T}
represents that the entry is the complex-conjugate of the pointed to entry.
"""
struct ConjPtr{T} 
    ptr::Ptr{T}
end

unsafe_convert(::Type{ConjPtr{T}}, Ac::Adjoint{<:Complex}) where T<:Complex = unsafe_convert(Ptr{T}, parent(Ac))    
function unsafe_convert(::Type{ConjPtr{T}}, V::SubArray{T,2}) where {T,N,P}
    kr, jr = parentindices(V)
    unsafe_convert(Ptr{T}, view(parent(V)', jr, kr))
end
```

File: stdlib/LinearAlgebra/src/adjtrans.jl
Comment (mbauman) [inline stdlib/LinearAlgebra/src/adjtrans.jl:206]: Here too.

Comment (dlfivefifty) [reply stdlib/LinearAlgebra/src/adjtrans.jl]: Fixed

**Comment by dlfivefifty** [comment]
Can this get merged so I can start testing all the packages I manage on Julia v1.5?

**Comment by KristofferC** [comment]
It seems a test was requested.

File: stdlib/LinearAlgebra/test/blas.jl
Comment (timholy) [inline stdlib/LinearAlgebra/test/blas.jl:468]: Worth a `@test_throws` for something like `strides(transpose(1:5))`?

**Comment by mbauman** [comment]
Good question... but is there a case in which it matters?  You're not allowed to multiply that first value by anything other than zero, are you?

Is there a way in which `(1, 1)` is any more correct or helpful than `(4, 1)`?

**Comment by dlfivefifty** [comment]
Only because `stride(::AbstractVector, 2)` is not defined at the moment. I changed the definition to work around this but if you really want `(4,1)` we can think of an alternative.

**Comment by mbauman** [comment]
This is my main motivation:

```Julia
julia> strides(randn(4, 1)')
(4, 1)

julia> strides(randn(4)')
(4, 1)
```

**Comment by dlfivefifty** [comment]
OK that's a strong argument. So I generalised `stride(::AbstractArray, k::Integer)` to be defined for all `k`, as in `Array`.

I also added `ConjPtr` to deal with pointers of complex adjoints. This could be removed if you aren't happy with it (though would mean ArrayLayouts.jl would still have some type piracy).

@mbauman Could you review the changes again please?

File: stdlib/LinearAlgebra/src/adjtrans.jl
Comment (KristofferC) [inline stdlib/LinearAlgebra/src/adjtrans.jl:216]: What issue does this fix? Since it has a docstring it looks like a new feature which we shouldn't put into 1.5.

Comment (dlfivefifty) [reply stdlib/LinearAlgebra/src/adjtrans.jl:216]: The issue is in the tests:
```julia
@test pointer(view(B',:,:)') == pointer(B)
```
I'll remove it though since it sounds like it will hold things up.

**Comment by dlfivefifty** [comment]
I believe this is ready to merge now unless you want other changes made.

File: base/abstractarray.jl
Comment (timholy) [inline base/abstractarray.jl:398]: So now `stride(1:5, 0)` and `stride(1:5, 2)` will return values but `stride(1:5, 1)` will throw an error? To me it seems better to throw a `MethodError` in all cases.

Comment (dlfivefifty) [reply base/abstractarray.jl:399]: I've modified it so it always calls `strides` now, producing the errors consistently now.

**Review by timholy — APPROVED** [review]

**Comment by KristofferC** [comment]
Whitespace needs fixing. If this is not merged soon, I'll just push for a 1.5-RC1 since the only thing this "fixes" is a package that was doing type piracy anyway and could just be worked around in the package.

File: base/abstractarray.jl
Comment (mbauman) [inline base/abstractarray.jl:397]: This feels funny to me; what is the meaning of `stride(A, 0)` or `stride(A, -1)`?  I see that this happens to be the behavior that falls out of the current `stride(::DenseArray, n)` implementation, but I'd rather keep the abstract behaviors more conservative and closer to the status quo  — especially to be considered for a back port.

File: base/abstractarray.jl
Comment (mbauman) [inline base/abstractarray.jl:399]: This doesn't seem to be the right answer, but I'm not sure what the right answer is.  Is it `sum(st .* size(A))`?

I know, it has a bit of that "does it matter what number is going to be multiplied by zero" rhetorical question quality, but this seems to be a better answer.

File: base/abstractarray.jl
Comment (dlfivefifty) [reply base/abstractarray.jl:397]: I agree, though then we shouldn't call `strides(A::AbstractVector, 2)` in the definition of `strides(A')`.

I could revert back to 
```julia
Base.strides(A::Adjoint{<:Real, <:AbstractVector}) = (1, stride(A.parent, 1))
```
despite it being inconsistent with your `strides(randn(4, 1)')` example (it's multipled by 0 anyways)

File: base/abstractarray.jl
Comment (mbauman) [reply base/abstractarray.jl:399]: Er, no, that's not quite it either.  Is it the absolute maximum of `st .* size(A)` (but preserving its sign)?  Even that leaves a bit to be desired as an abstract behavior... for example, here's how SubArray currently works:

```Julia
julia> v = @view rand(100)[1:3:end];

julia> stride(v, 1), stride(v, 2) # This is the answer the abstract behavior I proposed would give
(3, 102)

julia> m = @view rand(100, 1)[1:3:end, :];

julia> stride(m, 1), stride(m, 2) # This seems like a "better" answer:
(3, 100)
```

File: base/abstractarray.jl
Comment (mbauman) [reply base/abstractarray.jl:397]: There are two cases here, `n < 1` and `n > ndims(A)`.  This is only talking about the `n < 1` case — I see the motivation for the `n > ndims(A)` case.

Comment (dlfivefifty) [reply base/abstractarray.jl:397]: Ah ok (I don't see the motivation for either). So perhaps just delete `n < 1` case and let it throw an error?

Comment (mbauman) [reply base/abstractarray.jl:397]: Yes, that'd be my preference. :+1:

File: base/abstractarray.jl
Comment (mbauman) [reply base/abstractarray.jl:399]: I feel like an overestimate here is somehow more defensible than an underestimate as it pushes you _out of bounds_.  Here are a few more tricky n > ndims cases:

```Julia
julia> m  = @view rand(12, 24, 60)[1:3:end, end:-4:1, 2:-1:1];

julia> strides(m)
(3, -48, -288)

julia> stride(m, 4) # This seems reasonable; increasing this would jump out-of-bounds of the view
-576

julia> m2  = @view rand(12, 24, 60, 1)[1:3:end, end:-4:1, 2:-1:1, :];

julia> strides(m2) # this would _definitely_ jump out of bounds (that's the length of the parent, and we're starting in the middle!)
(3, -48, -288, 17280)

julia> p = PermutedDimsArray(m, (3,2,1));

julia> strides(p)
(-288, -48, 3)

julia> stride(p, 4)
ERROR: BoundsError: attempt to access (-288, -48, 3)
  at index [4]

julia> vpv = @view p[1:2, 1:2, 1:2]

julia> strides(vpv)
(-288, -48, 3)

julia> stride(vpv, 4) # I think this one is wrong; SubArray is implicitly assuming column-major parents
6
```

That last SubArray behavior can and should be tackled separately, but this gives me a bit of intuition here.

Comment (dlfivefifty) [reply base/abstractarray.jl:399]: Not sure I follow...

Comment (mbauman) [reply base/abstractarray.jl:399]: We're adding a behavior here — I just want to make sure that it makes sense with respect to our existing behaviors and that there's a decent rationale for it.

Comment (dlfivefifty) [reply base/abstractarray.jl:399]: Yes I get that, I don't follow what you are proposing. (It seems to me that it doesn't really matter what the stride is outside the range)

File: base/abstractarray.jl
Comment (JeffBezanson) [reply base/abstractarray.jl:397]: Bump

**Comment by dlfivefifty** [comment]
The test failure looks unrelated

**Comment by dlfivefifty** [comment]
Does "backport-1.5" mean it's not going to be in v1.5.0? If so I'll start working around the issue in ArrayLayouts.jl for v1.5-beta

**Comment by KristofferC** [comment]
> Does "backport-1.5" mean it's not going to be in v1.5.0?

No, if it gets merged before 1.5.0 is released, then it will be in.

File: base/abstractarray.jl
Comment (mbauman) [reply base/abstractarray.jl:398]: This was by-and-large just me talking out loud.  It's not a major thing since — as you say — it shouldn't matter.  Even if we use a value that's more likely to push someone out of bounds, we don't even check pointer accesses with `--check-bounds=yes`.  It _might_ be more likely that it'd help someone identify the problem, but that's quite the shot in the dark.

Upon reflection here I still do have a slight preference for `return sum(st .* size(A))` here simply because using `length` alone is a code smell in strides computations — it just looks wrong.  Indeed, it makes it clearer that we need to call `strides` in this branch (like Tim noted above).  But as a style thing, I'll leave the final decision to you.

Comment (dlfivefifty) [reply base/abstractarray.jl:398]: I'm happy with your suggestion .

Comment (mbauman) [reply base/abstractarray.jl:398]:
```suggestion
    return sum(st .* size(A))
```

**Review by mbauman — APPROVED** [review]

File: stdlib/LinearAlgebra/test/blas.jl
Comment (mbauman) [inline stdlib/LinearAlgebra/test/blas.jl:474]: Whoops, this needs to change too.

```suggestion
    @test stride(A,3) == stride(A,4) >= 6
```

**Pull request merged**
